File: rb/HostedRest_sdk.rb
# HostedRest SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'HostedRest_types'


class HostedRestSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = HostedRestUtility.new
    @_utility = utility

    config = HostedRestConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = HostedRestHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = HostedRestHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, HostedRestFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    HostedRestUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = HostedRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = HostedRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = HostedRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = HostedRestSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    # make_fetch_def returns a (fetchdef, err) tuple; destructure it and
    # return just the fetchdef Hash (raising on error) so callers — including
    # direct(), which indexes fetchdef["url"] — receive a Hash, mirroring the
    # ts/py prepare().
    fetchdef, fd_err = utility.make_fetch_def.call(ctx)
    raise fd_err if fd_err

    fetchdef
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue HostedRestError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = HostedRestHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = HostedRestHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Canonical facade: client.AgentHealth.list / client.AgentHealth.load({ "id" => ... })
  def AgentHealth(data = nil)
    require_relative 'entity/agent_health_entity'
    AgentHealthEntity.new(self, data)
  end


  # Canonical facade: client.AgentSandbox.list / client.AgentSandbox.load({ "id" => ... })
  def AgentSandbox(data = nil)
    require_relative 'entity/agent_sandbox_entity'
    AgentSandboxEntity.new(self, data)
  end


  # Canonical facade: client.AgentUserDetail.list / client.AgentUserDetail.load({ "id" => ... })
  def AgentUserDetail(data = nil)
    require_relative 'entity/agent_user_detail_entity'
    AgentUserDetailEntity.new(self, data)
  end


  # Canonical facade: client.AgentUserList.list / client.AgentUserList.load({ "id" => ... })
  def AgentUserList(data = nil)
    require_relative 'entity/agent_user_list_entity'
    AgentUserListEntity.new(self, data)
  end


  # Canonical facade: client.AppUser.list / client.AppUser.load({ "id" => ... })
  def AppUser(data = nil)
    require_relative 'entity/app_user_entity'
    AppUserEntity.new(self, data)
  end


  # Canonical facade: client.AppUserLogin.list / client.AppUserLogin.load({ "id" => ... })
  def AppUserLogin(data = nil)
    require_relative 'entity/app_user_login_entity'
    AppUserLoginEntity.new(self, data)
  end


  # Canonical facade: client.AppUserSession.list / client.AppUserSession.load({ "id" => ... })
  def AppUserSession(data = nil)
    require_relative 'entity/app_user_session_entity'
    AppUserSessionEntity.new(self, data)
  end


  # Canonical facade: client.AppUserTotal.list / client.AppUserTotal.load({ "id" => ... })
  def AppUserTotal(data = nil)
    require_relative 'entity/app_user_total_entity'
    AppUserTotalEntity.new(self, data)
  end


  # Canonical facade: client.AppUserVerify.list / client.AppUserVerify.load({ "id" => ... })
  def AppUserVerify(data = nil)
    require_relative 'entity/app_user_verify_entity'
    AppUserVerifyEntity.new(self, data)
  end


  # Canonical facade: client.Authentication.list / client.Authentication.load({ "id" => ... })
  def Authentication(data = nil)
    require_relative 'entity/authentication_entity'
    AuthenticationEntity.new(self, data)
  end


  # Canonical facade: client.Collection.list / client.Collection.load({ "id" => ... })
  def Collection(data = nil)
    require_relative 'entity/collection_entity'
    CollectionEntity.new(self, data)
  end


  # Canonical facade: client.CollectionRecord.list / client.CollectionRecord.load({ "id" => ... })
  def CollectionRecord(data = nil)
    require_relative 'entity/collection_record_entity'
    CollectionRecordEntity.new(self, data)
  end


  # Canonical facade: client.CollectionRecordList.list / client.CollectionRecordList.load({ "id" => ... })
  def CollectionRecordList(data = nil)
    require_relative 'entity/collection_record_list_entity'
    CollectionRecordListEntity.new(self, data)
  end


  # Canonical facade: client.Custom.list / client.Custom.load({ "id" => ... })
  def Custom(data = nil)
    require_relative 'entity/custom_entity'
    CustomEntity.new(self, data)
  end


  # Canonical facade: client.Legacy.list / client.Legacy.load({ "id" => ... })
  def Legacy(data = nil)
    require_relative 'entity/legacy_entity'
    LegacyEntity.new(self, data)
  end


  # Canonical facade: client.LegacyMutation.list / client.LegacyMutation.load({ "id" => ... })
  def LegacyMutation(data = nil)
    require_relative 'entity/legacy_mutation_entity'
    LegacyMutationEntity.new(self, data)
  end


  # Canonical facade: client.LegacyUnknown.list / client.LegacyUnknown.load({ "id" => ... })
  def LegacyUnknown(data = nil)
    require_relative 'entity/legacy_unknown_entity'
    LegacyUnknownEntity.new(self, data)
  end


  # Canonical facade: client.LegacyUnknownList.list / client.LegacyUnknownList.load({ "id" => ... })
  def LegacyUnknownList(data = nil)
    require_relative 'entity/legacy_unknown_list_entity'
    LegacyUnknownListEntity.new(self, data)
  end


  # Canonical facade: client.LegacyUser.list / client.LegacyUser.load({ "id" => ... })
  def LegacyUser(data = nil)
    require_relative 'entity/legacy_user_entity'
    LegacyUserEntity.new(self, data)
  end


  # Canonical facade: client.LegacyUserList.list / client.LegacyUserList.load({ "id" => ... })
  def LegacyUserList(data = nil)
    require_relative 'entity/legacy_user_list_entity'
    LegacyUserListEntity.new(self, data)
  end


  # Canonical facade: client.Login.list / client.Login.load({ "id" => ... })
  def Login(data = nil)
    require_relative 'entity/login_entity'
    LoginEntity.new(self, data)
  end


  # Canonical facade: client.Register.list / client.Register.load({ "id" => ... })
  def Register(data = nil)
    require_relative 'entity/register_entity'
    RegisterEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = HostedRestSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: py/hostedrest_sdk.py
# HostedRest SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import HostedRestUtility
from core.spec import HostedRestSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import HostedRestBaseFeature
from features import _make_feature


class HostedRestSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = HostedRestUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return HostedRestUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = HostedRestSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def agent_health(self):
        """Idiomatic facade: client.agent_health.list() / client.agent_health.load({"id": ...})."""
        from entity.agent_health_entity import AgentHealthEntity
        cached = getattr(self, "_agent_health", None)
        if cached is None:
            cached = AgentHealthEntity(self, None)
            self._agent_health = cached
        return cached

    def AgentHealth(self, data=None):
        # Deprecated: use client.agent_health instead.
        from entity.agent_health_entity import AgentHealthEntity
        return AgentHealthEntity(self, data)


    @property
    def agent_sandbox(self):
        """Idiomatic facade: client.agent_sandbox.list() / client.agent_sandbox.load({"id": ...})."""
        from entity.agent_sandbox_entity import AgentSandboxEntity
        cached = getattr(self, "_agent_sandbox", None)
        if cached is None:
            cached = AgentSandboxEntity(self, None)
            self._agent_sandbox = cached
        return cached

    def AgentSandbox(self, data=None):
        # Deprecated: use client.agent_sandbox instead.
        from entity.agent_sandbox_entity import AgentSandboxEntity
        return AgentSandboxEntity(self, data)


    @property
    def agent_user_detail(self):
        """Idiomatic facade: client.agent_user_detail.list() / client.agent_user_detail.load({"id": ...})."""
        from entity.agent_user_detail_entity import AgentUserDetailEntity
        cached = getattr(self, "_agent_user_detail", None)
        if cached is None:
            cached = AgentUserDetailEntity(self, None)
            self._agent_user_detail = cached
        return cached

    def AgentUserDetail(self, data=None):
        # Deprecated: use client.agent_user_detail instead.
        from entity.agent_user_detail_entity import AgentUserDetailEntity
        return AgentUserDetailEntity(self, data)


    @property
    def agent_user_list(self):
        """Idiomatic facade: client.agent_user_list.list() / client.agent_user_list.load({"id": ...})."""
        from entity.agent_user_list_entity import AgentUserListEntity
        cached = getattr(self, "_agent_user_list", None)
        if cached is None:
            cached = AgentUserListEntity(self, None)
            self._agent_user_list = cached
        return cached

    def AgentUserList(self, data=None):
        # Deprecated: use client.agent_user_list instead.
        from entity.agent_user_list_entity import AgentUserListEntity
        return AgentUserListEntity(self, data)


    @property
    def app_user(self):
        """Idiomatic facade: client.app_user.list() / client.app_user.load({"id": ...})."""
        from entity.app_user_entity import AppUserEntity
        cached = getattr(self, "_app_user", None)
        if cached is None:
            cached = AppUserEntity(self, None)
            self._app_user = cached
        return cached

    def AppUser(self, data=None):
        # Deprecated: use client.app_user instead.
        from entity.app_user_entity import AppUserEntity
        return AppUserEntity(self, data)


    @property
    def app_user_login(self):
        """Idiomatic facade: client.app_user_login.list() / client.app_user_login.load({"id": ...})."""
        from entity.app_user_login_entity import AppUserLoginEntity
        cached = getattr(self, "_app_user_login", None)
        if cached is None:
            cached = AppUserLoginEntity(self, None)
            self._app_user_login = cached
        return cached

    def AppUserLogin(self, data=None):
        # Deprecated: use client.app_user_login instead.
        from entity.app_user_login_entity import AppUserLoginEntity
        return AppUserLoginEntity(self, data)


    @property
    def app_user_session(self):
        """Idiomatic facade: client.app_user_session.list() / client.app_user_session.load({"id": ...})."""
        from entity.app_user_session_entity import AppUserSessionEntity
        cached = getattr(self, "_app_user_session", None)
        if cached is None:
            cached = AppUserSessionEntity(self, None)
            self._app_user_session = cached
        return cached

    def AppUserSession(self, data=None):
        # Deprecated: use client.app_user_session instead.
        from entity.app_user_session_entity import AppUserSessionEntity
        return AppUserSessionEntity(self, data)


    @property
    def app_user_total(self):
        """Idiomatic facade: client.app_user_total.list() / client.app_user_total.load({"id": ...})."""
        from entity.app_user_total_entity import AppUserTotalEntity
        cached = getattr(self, "_app_user_total", None)
        if cached is None:
            cached = AppUserTotalEntity(self, None)
            self._app_user_total = cached
        return cached

    def AppUserTotal(self, data=None):
        # Deprecated: use client.app_user_total instead.
        from entity.app_user_total_entity import AppUserTotalEntity
        return AppUserTotalEntity(self, data)


    @property
    def app_user_verify(self):
        """Idiomatic facade: client.app_user_verify.list() / client.app_user_verify.load({"id": ...})."""
        from entity.app_user_verify_entity import AppUserVerifyEntity
        cached = getattr(self, "_app_user_verify", None)
        if cached is None:
            cached = AppUserVerifyEntity(self, None)
            self._app_user_verify = cached
        return cached

    def AppUserVerify(self, data=None):
        # Deprecated: use client.app_user_verify instead.
        from entity.app_user_verify_entity import AppUserVerifyEntity
        return AppUserVerifyEntity(self, data)


    @property
    def authentication(self):
        """Idiomatic facade: client.authentication.list() / client.authentication.load({"id": ...})."""
        from entity.authentication_entity import AuthenticationEntity
        cached = getattr(self, "_authentication", None)
        if cached is None:
            cached = AuthenticationEntity(self, None)
            self._authentication = cached
        return cached

    def Authentication(self, data=None):
        # Deprecated: use client.authentication instead.
        from entity.authentication_entity import AuthenticationEntity
        return AuthenticationEntity(self, data)


    @property
    def collection(self):
        """Idiomatic facade: client.collection.list() / client.collection.load({"id": ...})."""
        from entity.collection_entity import CollectionEntity
        cached = getattr(self, "_collection", None)
        if cached is None:
            cached = CollectionEntity(self, None)
            self._collection = cached
        return cached

    def Collection(self, data=None):
        # Deprecated: use client.collection instead.
        from entity.collection_entity import CollectionEntity
        return CollectionEntity(self, data)


    @property
    def collection_record(self):
        """Idiomatic facade: client.collection_record.list() / client.collection_record.load({"id": ...})."""
        from entity.collection_record_entity import CollectionRecordEntity
        cached = getattr(self, "_collection_record", None)
        if cached is None:
            cached = CollectionRecordEntity(self, None)
            self._collection_record = cached
        return cached

    def CollectionRecord(self, data=None):
        # Deprecated: use client.collection_record instead.
        from entity.collection_record_entity import CollectionRecordEntity
        return CollectionRecordEntity(self, data)


    @property
    def collection_record_list(self):
        """Idiomatic facade: client.collection_record_list.list() / client.collection_record_list.load({"id": ...})."""
        from entity.collection_record_list_entity import CollectionRecordListEntity
        cached = getattr(self, "_collection_record_list", None)
        if cached is None:
            cached = CollectionRecordListEntity(self, None)
            self._collection_record_list = cached
        return cached

    def CollectionRecordList(self, data=None):
        # Deprecated: use client.collection_record_list instead.
        from entity.collection_record_list_entity import CollectionRecordListEntity
        return CollectionRecordListEntity(self, data)


    @property
    def custom(self):
        """Idiomatic facade: client.custom.list() / client.custom.load({"id": ...})."""
        from entity.custom_entity import CustomEntity
        cached = getattr(self, "_custom", None)
        if cached is None:
            cached = CustomEntity(self, None)
            self._custom = cached
        return cached

    def Custom(self, data=None):
        # Deprecated: use client.custom instead.
        from entity.custom_entity import CustomEntity
        return CustomEntity(self, data)


    @property
    def legacy(self):
        """Idiomatic facade: client.legacy.list() / client.legacy.load({"id": ...})."""
        from entity.legacy_entity import LegacyEntity
        cached = getattr(self, "_legacy", None)
        if cached is None:
            cached = LegacyEntity(self, None)
            self._legacy = cached
        return cached

    def Legacy(self, data=None):
        # Deprecated: use client.legacy instead.
        from entity.legacy_entity import LegacyEntity
        return LegacyEntity(self, data)


    @property
    def legacy_mutation(self):
        """Idiomatic facade: client.legacy_mutation.list() / client.legacy_mutation.load({"id": ...})."""
        from entity.legacy_mutation_entity import LegacyMutationEntity
        cached = getattr(self, "_legacy_mutation", None)
        if cached is None:
            cached = LegacyMutationEntity(self, None)
            self._legacy_mutation = cached
        return cached

    def LegacyMutation(self, data=None):
        # Deprecated: use client.legacy_mutation instead.
        from entity.legacy_mutation_entity import LegacyMutationEntity
        return LegacyMutationEntity(self, data)


    @property
    def legacy_unknown(self):
        """Idiomatic facade: client.legacy_unknown.list() / client.legacy_unknown.load({"id": ...})."""
        from entity.legacy_unknown_entity import LegacyUnknownEntity
        cached = getattr(self, "_legacy_unknown", None)
        if cached is None:
            cached = LegacyUnknownEntity(self, None)
            self._legacy_unknown = cached
        return cached

    def LegacyUnknown(self, data=None):
        # Deprecated: use client.legacy_unknown instead.
        from entity.legacy_unknown_entity import LegacyUnknownEntity
        return LegacyUnknownEntity(self, data)


    @property
    def legacy_unknown_list(self):
        """Idiomatic facade: client.legacy_unknown_list.list() / client.legacy_unknown_list.load({"id": ...})."""
        from entity.legacy_unknown_list_entity import LegacyUnknownListEntity
        cached = getattr(self, "_legacy_unknown_list", None)
        if cached is None:
            cached = LegacyUnknownListEntity(self, None)
            self._legacy_unknown_list = cached
        return cached

    def LegacyUnknownList(self, data=None):
        # Deprecated: use client.legacy_unknown_list instead.
        from entity.legacy_unknown_list_entity import LegacyUnknownListEntity
        return LegacyUnknownListEntity(self, data)


    @property
    def legacy_user(self):
        """Idiomatic facade: client.legacy_user.list() / client.legacy_user.load({"id": ...})."""
        from entity.legacy_user_entity import LegacyUserEntity
        cached = getattr(self, "_legacy_user", None)
        if cached is None:
            cached = LegacyUserEntity(self, None)
            self._legacy_user = cached
        return cached

    def LegacyUser(self, data=None):
        # Deprecated: use client.legacy_user instead.
        from entity.legacy_user_entity import LegacyUserEntity
        return LegacyUserEntity(self, data)


    @property
    def legacy_user_list(self):
        """Idiomatic facade: client.legacy_user_list.list() / client.legacy_user_list.load({"id": ...})."""
        from entity.legacy_user_list_entity import LegacyUserListEntity
        cached = getattr(self, "_legacy_user_list", None)
        if cached is None:
            cached = LegacyUserListEntity(self, None)
            self._legacy_user_list = cached
        return cached

    def LegacyUserList(self, data=None):
        # Deprecated: use client.legacy_user_list instead.
        from entity.legacy_user_list_entity import LegacyUserListEntity
        return LegacyUserListEntity(self, data)


    @property
    def login(self):
        """Idiomatic facade: client.login.list() / client.login.load({"id": ...})."""
        from entity.login_entity import LoginEntity
        cached = getattr(self, "_login", None)
        if cached is None:
            cached = LoginEntity(self, None)
            self._login = cached
        return cached

    def Login(self, data=None):
        # Deprecated: use client.login instead.
        from entity.login_entity import LoginEntity
        return LoginEntity(self, data)


    @property
    def register(self):
        """Idiomatic facade: client.register.list() / client.register.load({"id": ...})."""
        from entity.register_entity import RegisterEntity
        cached = getattr(self, "_register", None)
        if cached is None:
            cached = RegisterEntity(self, None)
            self._register = cached
        return cached

    def Register(self, data=None):
        # Deprecated: use client.register instead.
        from entity.register_entity import RegisterEntity
        return RegisterEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk
